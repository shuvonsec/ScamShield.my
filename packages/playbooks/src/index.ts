export type PlaybookChange = {
  description: string;
  file?: string;
  diff?: string;
};

export type Playbook = {
  id: string;
  name: string;
  description: string;
  risks: string[];
  apply: (params: Record<string, unknown>) => Promise<PlaybookChange[]>;
  rollback: (params: Record<string, unknown>) => Promise<PlaybookChange[]>;
};

function delay(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function noopRollback() {
  return [{ description: 'Rollback pending implementation.' }];
}

export const library: Playbook[] = [
  {
    id: 'wordpress-hardening',
    name: 'WordPress Hardening Essentials',
    description: 'Disable XML-RPC, enforce updates, and audit plugins.',
    risks: ['Brute force', 'Plugin abuse', 'XML-RPC amplification'],
    async apply() {
      await delay(50);
      return [
        { description: 'Disable xmlrpc.php via .htaccess rule', file: '.htaccess', diff: '-<Files xmlrpc.php>\n+deny from all' },
        { description: 'Enable automatic core and plugin updates', file: 'wp-config.php' },
        { description: 'Generate plugin audit report', file: 'reports/plugin-audit.json' }
      ];
    },
    rollback: noopRollback
  },
  {
    id: 'laravel-nginx',
    name: 'Laravel & Nginx Security Headers',
    description: 'Applies security headers, rate limits, and blocks sensitive files.',
    risks: ['Missing security headers', 'Exposure of .env', 'Credential stuffing'],
    async apply() {
      await delay(50);
      return [
        { description: 'Add security headers middleware', file: 'app/Http/Middleware/SecureHeaders.php' },
        { description: 'Update nginx.conf with rate limiting', file: 'deploy/nginx.conf' },
        { description: 'Block /.env and /.git locations', file: 'deploy/nginx.conf' }
      ];
    },
    rollback: noopRollback
  },
  {
    id: 'cloudflare-waf',
    name: 'Cloudflare WAF Shield',
    description: 'Creates WAF custom rules for login protection and scam blocking.',
    risks: ['Credential stuffing', 'Phishing pages', 'Bot attacks'],
    async apply(params) {
      await delay(50);
      return [
        { description: 'Create firewall rule to challenge suspicious login attempts', diff: JSON.stringify(params) },
        { description: 'Enable bot fight mode', diff: 'bot_fight_mode: on' }
      ];
    },
    rollback: noopRollback
  },
  {
    id: 'csp-builder',
    name: 'Content Security Policy Builder',
    description: 'Generates a CSP with report-only toggle.',
    risks: ['XSS', 'Malicious script injection'],
    async apply(params) {
      const reportOnly = Boolean(params.reportOnly);
      await delay(10);
      return [
        {
          description: `Apply CSP header (${reportOnly ? 'report-only' : 'enforce'})`,
          diff: `Content-Security-Policy${reportOnly ? '-Report-Only' : ''}: default-src 'self'; img-src 'self' data:; connect-src 'self';`
        }
      ];
    },
    rollback: noopRollback
  }
];
