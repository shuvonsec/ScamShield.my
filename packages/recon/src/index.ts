export type ReconFinding = {
  severity: 'critical' | 'high' | 'medium' | 'low';
  category: string;
  title: string;
  evidence: Record<string, unknown>;
  fix: Record<string, unknown>;
};

export type ReconResult = {
  subdomains: string[];
  findings: ReconFinding[];
};

const COMMON_SUBDOMAINS = ['www', 'admin', 'portal', 'vpn', 'mail'];

export async function runRecon(domain: string): Promise<ReconResult> {
  const subdomains = await enumerateSubdomains(domain);
  const probes = await Promise.all(subdomains.map((host) => probeHttp(host)));
  const findings: ReconFinding[] = [];

  for (const probe of probes) {
    findings.push(...auditSecurityHeaders(probe));
    findings.push(...detectCmsIssues(probe));
  }

  return {
    subdomains,
    findings
  };
}

export async function enumerateSubdomains(domain: string) {
  const generated = COMMON_SUBDOMAINS.map((prefix) => `${prefix}.${domain}`);
  return [domain, ...generated];
}

type ProbeResult = {
  host: string;
  status: number;
  headers: Record<string, string>;
  technologies: string[];
};

export async function probeHttp(host: string): Promise<ProbeResult> {
  try {
    const response = await fetch(`https://${host}`, { method: 'HEAD' });
    return {
      host,
      status: response.status,
      headers: Object.fromEntries(response.headers.entries()),
      technologies: fingerprint(response.headers)
    };
  } catch (error) {
    return {
      host,
      status: 0,
      headers: {},
      technologies: []
    };
  }
}

function fingerprint(headers: Headers) {
  const tech: string[] = [];
  if (headers.get('server')?.toLowerCase().includes('nginx')) tech.push('nginx');
  if (headers.get('x-powered-by')?.toLowerCase().includes('php')) tech.push('php');
  if (headers.get('x-powered-by')?.toLowerCase().includes('express')) tech.push('express');
  return tech;
}

function auditSecurityHeaders(probe: ProbeResult): ReconFinding[] {
  const findings: ReconFinding[] = [];
  const headers = probe.headers;

  const requiredHeaders = ['content-security-policy', 'strict-transport-security', 'x-frame-options'];
  for (const key of requiredHeaders) {
    if (!headers[key]) {
      findings.push({
        severity: 'medium',
        category: 'headers',
        title: `${probe.host} missing ${key}`,
        evidence: { host: probe.host, header: key },
        fix: { recommendation: `Apply security header ${key}` }
      });
    }
  }

  return findings;
}

function detectCmsIssues(probe: ProbeResult): ReconFinding[] {
  const findings: ReconFinding[] = [];
  if (probe.technologies.includes('php')) {
    findings.push({
      severity: 'high',
      category: 'cms',
      title: `${probe.host} appears to run PHP without hardened headers`,
      evidence: { host: probe.host, technologies: probe.technologies },
      fix: { playbook: 'laravel-nginx' }
    });
  }
  if (probe.headers['x-generator']?.toLowerCase().includes('wordpress')) {
    findings.push({
      severity: 'critical',
      category: 'wordpress',
      title: `${probe.host} WordPress installation missing XML-RPC hardening`,
      evidence: { host: probe.host },
      fix: { playbook: 'wordpress-hardening' }
    });
  }
  return findings;
}
