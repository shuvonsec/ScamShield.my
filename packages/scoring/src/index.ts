export type VerdictLabel = 'safe' | 'unknown' | 'suspicious' | 'malicious';

export type ScoreResult = {
  score: number;
  label: VerdictLabel;
  signals: string[];
};

const suspiciousTokens = [
  'login',
  'verify',
  'reset',
  'bank',
  'wallet',
  'invoice',
  'bonus',
  'gift',
  'prize'
];

const riskyTlds = ['.zip', '.mov', '.rest', '.quest', '.country'];

export function scoreUrl(rawUrl: string): ScoreResult {
  const url = new URL(rawUrl);
  const signals: string[] = [];
  let score = 0;

  if (url.hostname.includes('xn--')) {
    score += 20;
    signals.push('URL contains punycode/IDN characters.');
  }

  if (url.href.length > 80) {
    score += 10;
    signals.push('URL length is unusually long.');
  }

  const path = `${url.pathname}${url.search}${url.hash}`;
  const digits = (url.hostname + path).replace(/\D/g, '').length;
  const letters = (url.hostname + path).replace(/[^a-zA-Z]/g, '').length;
  if (digits > 0 && digits / Math.max(letters, 1) > 0.5) {
    score += 10;
    signals.push('High digit-to-letter ratio indicative of obfuscation.');
  }

  for (const token of suspiciousTokens) {
    if (url.href.toLowerCase().includes(token)) {
      score += 8;
      signals.push(`Contains high-risk keyword: ${token}`);
    }
  }

  if (url.pathname.includes('@')) {
    score += 15;
    signals.push('Contains @ character in path which may hide real destination.');
  }

  const subdomainCount = url.hostname.split('.').length - 2;
  if (subdomainCount > 2) {
    score += 10;
    signals.push('Multiple nested subdomains detected.');
  }

  if (riskyTlds.some((tld) => url.hostname.endsWith(tld))) {
    score += 15;
    signals.push('Domain ends with a risky or abuse-prone TLD.');
  }

  if (!['http:', 'https:'].includes(url.protocol)) {
    score += 5;
    signals.push('Non-HTTP protocol detected.');
  }

  if (url.searchParams.toString().length > 100) {
    score += 5;
    signals.push('Query string is unusually long.');
  }

  score = Math.min(100, score);

  let label: VerdictLabel = 'safe';
  if (score >= 70) {
    label = 'malicious';
  } else if (score >= 40) {
    label = 'suspicious';
  } else if (score >= 20) {
    label = 'unknown';
  }

  return { score, label, signals };
}

export function scoreText(text: string): ScoreResult {
  const urls = text.match(/https?:\/\/[^\s]+/g);
  if (!urls) {
    return {
      score: 10,
      label: 'unknown',
      signals: ['No URL detected. Assess the sender before acting.']
    };
  }
  const urlScore = scoreUrl(urls[0]);
  return {
    ...urlScore,
    signals: ['Extracted URL from message.', ...urlScore.signals]
  };
}
