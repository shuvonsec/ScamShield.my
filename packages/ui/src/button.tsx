import { Slot } from '@radix-ui/react-slot';
import { type ButtonHTMLAttributes, forwardRef } from 'react';
import { cn } from './cn';

type ButtonVariant = 'default' | 'outline' | 'ghost' | 'secondary';

type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> & {
  variant?: ButtonVariant;
  asChild?: boolean;
};

const variants: Record<ButtonVariant, string> = {
  default: 'bg-blue-600 text-white hover:bg-blue-500',
  outline: 'border border-slate-700 text-slate-100 hover:bg-slate-900',
  ghost: 'text-slate-200 hover:bg-slate-900',
  secondary: 'bg-slate-700 text-white hover:bg-slate-600'
};

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  ({ className, variant = 'default', asChild = false, ...props }, ref) => {
    const Comp = asChild ? Slot : 'button';
    return (
      <Comp
        ref={ref as never}
        className={cn(
          'inline-flex h-10 items-center justify-center rounded-md px-4 py-2 text-sm font-medium transition-colors focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2 focus:ring-offset-slate-950',
          variants[variant],
          className
        )}
        {...props}
      />
    );
  }
);

Button.displayName = 'Button';
