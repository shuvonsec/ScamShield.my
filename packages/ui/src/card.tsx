import { cn } from './cn';
import { type HTMLAttributes } from 'react';

export function Card({ className, ...props }: HTMLAttributes<HTMLDivElement>) {
  return <div className={cn('rounded-lg border border-slate-800 bg-slate-900 p-4', className)} {...props} />;
}

export function CardHeader({ className, ...props }: HTMLAttributes<HTMLDivElement>) {
  return <div className={cn('space-y-1.5', className)} {...props} />;
}

export function CardTitle({ className, ...props }: HTMLAttributes<HTMLHeadingElement>) {
  return <h3 className={cn('text-lg font-semibold leading-none tracking-tight', className)} {...props} />;
}

export function CardContent({ className, ...props }: HTMLAttributes<HTMLDivElement>) {
  return <div className={cn('pt-2', className)} {...props} />;
}

export function CardDescription({ className, ...props }: HTMLAttributes<HTMLParagraphElement>) {
  return <p className={cn('text-sm text-slate-400', className)} {...props} />;
}
