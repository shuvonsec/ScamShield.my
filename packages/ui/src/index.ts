export * from './components/button';
export * from './components/card';
export * from './components/input';
export * from './components/textarea';
export * from './components/label';
export * from './components/badge';
export * from './components/tabs';
export * from './components/table';
export * from './components/dialog';
export * from './components/select';
export * from './components/accordion';
export * from './utils/cn';
