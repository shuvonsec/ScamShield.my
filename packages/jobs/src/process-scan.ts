import type { PrismaClient } from '@prisma/client';
import { runRecon } from '@scamshield/recon';
import { library } from '@scamshield/playbooks';

export type ProcessScanOptions = {
  prisma: PrismaClient;
  scanId: string;
};

export async function processScan({ prisma, scanId }: ProcessScanOptions) {
  const scan = await prisma.scan.findUnique({
    where: { id: scanId },
    include: { project: true }
  });

  if (!scan) {
    throw new Error(`Scan ${scanId} not found`);
  }

  await prisma.scan.update({
    where: { id: scan.id },
    data: { status: 'running' }
  });

  const recon = await runRecon(scan.project.domain);

  for (const finding of recon.findings) {
    await prisma.finding.create({
      data: {
        scanId: scan.id,
        severity: finding.severity,
        category: finding.category,
        title: finding.title,
        evidence: finding.evidence,
        fix: finding.fix,
        status: 'open'
      }
    });
  }

  await prisma.scan.update({
    where: { id: scan.id },
    data: {
      status: 'completed',
      finishedAt: new Date()
    }
  });

  return {
    findings: recon.findings.length,
    recommendedPlaybooks: library.map((playbook) => ({
      id: playbook.id,
      name: playbook.name
    }))
  };
}
