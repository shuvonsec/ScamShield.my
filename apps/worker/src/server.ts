import http from 'http';
import { processScan } from './handlers/process-scan';

const server = http.createServer(async (req, res) => {
  if (req.method === 'POST' && req.url === '/process-scan') {
    let body = '';
    req.on('data', (chunk) => {
      body += chunk;
    });
    req.on('end', async () => {
      const data = JSON.parse(body);
      const result = await processScan(data.scanId);
      res.writeHead(200, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify(result));
    });
  } else {
    res.writeHead(404);
    res.end();
  }
});

const port = process.env.PORT || 4000;
server.listen(port, () => {
  console.log(`Worker listening on ${port}`);
});
