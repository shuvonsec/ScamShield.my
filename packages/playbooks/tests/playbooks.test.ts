import { describe, expect, it } from 'vitest';
import { library } from '../src';

describe('playbooks', () => {
  it('generates deterministic changes', async () => {
    const playbook = library.find((p) => p.id === 'csp-builder');
    expect(playbook).toBeTruthy();
    const changes = await playbook!.apply({ reportOnly: true });
    expect(changes).toMatchSnapshot();
  });
});
