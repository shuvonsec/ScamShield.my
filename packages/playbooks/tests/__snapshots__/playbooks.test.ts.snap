// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`playbooks generates deterministic changes 1`] = `[
  {
    "description": "Apply CSP header (report-only)",
    "diff": "Content-Security-Policy-Report-Only: default-src 'self'; img-src 'self' data:; connect-src 'self';"
  }
]`;
